### ISSUE TITLE
Showing wrong compiler error in this case.

### ISSUE THREAD

**Issue by 623637646** [rust-lang/rust#139680]
### Code

```Rust
// If we remove this enum, the compiler error seems reasonable.
enum Terminal<E> {
    Error(E),
}

struct From<I>(I);

fn repeat<T>(value: T, n: usize) -> From<_>
where
    T: Clone,
{
    From(std::iter::repeat_n(value, n))
}
```

### Current output

```Shell
[{
	"resource": "/Users/yawang/Documents/rust-test/src/lib.rs",
	"owner": "rustc",
	"code": {
		"value": "Click for full compiler diagnostic",
		"target": {
			"$mid": 1,
			"path": "/diagnostic message [0]",
			"scheme": "rust-analyzer-diagnostics-view",
			"query": "0",
			"fragment": "file:///Users/yawang/Documents/rust-test/src/lib.rs"
		}
	},
	"severity": 8,
	"message": "cycle detected when computing the variances for items in this crate\n...which again requires computing the variances for items in this crate, completing the cycle\nsee https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information",
	"source": "rustc",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 1,
	"relatedInformation": [
		{
			"startLineNumber": 7,
			"startColumn": 1,
			"endLineNumber": 9,
			"endColumn": 14,
			"message": "...which requires computing function signature of `repeat`...",
			"resource": "/Users/yawang/Documents/rust-test/src/lib.rs"
		},
		{
			"startLineNumber": 7,
			"startColumn": 1,
			"endLineNumber": 9,
			"endColumn": 14,
			"message": "...which requires type-checking `repeat`...",
			"resource": "/Users/yawang/Documents/rust-test/src/lib.rs"
		},
		{
			"startLineNumber": 5,
			"startColumn": 1,
			"endLineNumber": 5,
			"endColumn": 15,
			"message": "...which requires computing the variances of `From`...",
			"resource": "/Users/yawang/Documents/rust-test/src/lib.rs"
		},
		{
			"startLineNumber": 1,
			"startColumn": 1,
			"endLineNumber": 1,
			"endColumn": 17,
			"message": "cycle used when computing the variances of `Terminal`",
			"resource": "/Users/yawang/Documents/rust-test/src/lib.rs"
		}
	]
}]
```

### Desired output

```Shell
[{
	"resource": "/Users/yawang/Documents/rust-test/src/lib.rs",
	"owner": "rustc",
	"code": {
		"value": "Click for full compiler diagnostic",
		"target": {
			"$mid": 1,
			"path": "/diagnostic message [1]",
			"scheme": "rust-analyzer-diagnostics-view",
			"query": "1",
			"fragment": "file:///Users/yawang/Documents/rust-test/src/lib.rs"
		}
	},
	"severity": 8,
	"message": "the placeholder `_` is not allowed within types on item signatures for return types\nnot allowed in type signatures",
	"source": "rustc",
	"startLineNumber": 7,
	"startColumn": 42,
	"endLineNumber": 7,
	"endColumn": 43
},{
	"resource": "/Users/yawang/Documents/rust-test/src/lib.rs",
	"owner": "rustc",
	"code": {
		"value": "Click for full compiler diagnostic",
		"target": {
			"$mid": 1,
			"path": "/diagnostic message [0]",
			"scheme": "rust-analyzer-diagnostics-view",
			"query": "0",
			"fragment": "file:///Users/yawang/Documents/rust-test/src/lib.rs"
		}
	},
	"severity": 8,
	"message": "type annotations needed\ncannot infer type for struct `From<_>`",
	"source": "rustc",
	"startLineNumber": 10,
	"startColumn": 1,
	"endLineNumber": 12,
	"endColumn": 2
}]
```

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: aarch64-apple-darwin
release: 1.86.0
LLVM version: 19.1.7
```

**Comment by shadyfennec** [comment]
I just hit a similar error with this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=4cd322ea6ea8c397c5121680c434313d)):

```rust
pub struct Map;

pub struct View<'a> {
    map: &'a Map,
}

impl<'a> View<'a> {
    pub fn iter<'r>(&'r self) -> Iter<'a, 'r, _> {
        
        Iter {
            view: self,
            inner: None
        }
    }
}

pub struct Iter<'a, 'r, I> {
    view: &'r View<'a>,
    inner: Option<I>
}
```

The error displayed:

```
Compiling playground v0.0.1 (/playground)
error[E0391]: cycle detected when computing the variances for items in this crate
   |
note: ...which requires computing function signature of `<impl at src/lib.rs:7:1: 7:18>::iter`...
  --> src/lib.rs:8:5
   |
8  |     pub fn iter<'r>(&'r self) -> Iter<'a, 'r, _> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires type-checking `<impl at src/lib.rs:7:1: 7:18>::iter`...
  --> src/lib.rs:8:5
   |
8  |     pub fn iter<'r>(&'r self) -> Iter<'a, 'r, _> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing the variances of `Iter`...
  --> src/lib.rs:17:1
   |
17 | pub struct Iter<'a, 'r, I> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: ...which again requires computing the variances for items in this crate, completing the cycle
note: cycle used when computing the variances of `View`
  --> src/lib.rs:3:1
   |
3  | pub struct View<'a> {
   | ^^^^^^^^^^^^^^^^^^^
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

For more information about this error, try `rustc --explain E0391`.
error: could not compile `playground` (lib) due to 1 previous error
```

Replacing the type placeholder for the return type of `iter()` by a concrete type or a generic solves the error. I'm not sure that the error, "`cycle detected when computing the variances for items in this crate`", is related to the actual problem. The actual problem seems more related to the error written in the desired output of this issue, relating to the use of a placeholder in the return type.
